Core/src/Vertexing/AdaptiveGridTrackDensity.cpp

```diff
@@ -75,197 +91,171 @@ Acts::AdaptiveGridTrackDensity::getMaxZTPositionAndWidth(
   if (!widthRes.ok()) {
     return widthRes.error();
   }
-  float width = *widthRes;
+  double width = *widthRes;
   ZTPositionAndWidth maxZTAndWidth{maxZT, width};
   return maxZTAndWidth;
 }
 
-typename Acts::AdaptiveGridTrackDensity::DensityMap
-Acts::AdaptiveGridTrackDensity::addTrack(const Acts::BoundTrackParameters& trk,
-                                         DensityMap& mainDensityMap) const {
+AdaptiveGridTrackDensity::DensityMap AdaptiveGridTrackDensity::addTrack(
+    const BoundTrackParameters& trk, DensityMap& mainDensityMap) const {
   ActsVector<3> impactParams = trk.impactParameters();
   ActsSquareMatrix<3> cov = trk.impactParameterCovariance().value();
 
-  // Calculate bin in d direction
-  int centralDBin = getBin(impactParams(0), m_cfg.spatialBinExtent);
-  // Check if current track affects grid density
-  if (std::abs(centralDBin) > (m_cfg.spatialTrkGridSize - 1) / 2.) {
-    DensityMap emptyTrackDensityMap;
-    return emptyTrackDensityMap;
-  }
-  // Calculate bin in z direction
-  int centralZBin = getBin(impactParams(1), m_cfg.spatialBinExtent);
+  int spatialTrkGridSize = getSpatialTrkGridSize(std::sqrt(cov(1, 1)));
+  int temporalTrkGridSize = getTemporalTrkGridSize(std::sqrt(cov(2, 2)));
 
-  // If we don't do time vertex seeding, the time index is set to 0
-  Bin centralBin = std::make_pair(centralZBin, 0.);
+  // Calculate bin in z and t direction
+  int centralZBin = getBin(impactParams(1), m_cfg.spatialBinExtent);
+  int centralTBin = getBin(impactParams(2), m_cfg.temporalBinExtent);
 
-  // Calculate bin in t direction if we do time vertex seeding
-  if (m_cfg.temporalTrkGridSize > 1) {
-    int centralTBin = getBin(impactParams(2), m_cfg.temporalBinExtent);
-    centralBin.second = centralTBin;
-  }
+  Bin centralBin = {centralZBin, centralTBin};
 
-  DensityMap trackDensityMap = createTrackGrid(impactParams, centralBin, cov);
+  DensityMap trackDensityMap = createTrackGrid(
+      impactParams, centralBin, cov, spatialTrkGridSize, temporalTrkGridSize);
 
-  for (const auto& densityEntry : trackDensityMap) {
-    Bin bin = densityEntry.first;
-    float trackDensity = densityEntry.second;
-    // Check if z bin is already part of the main grid
-    if (mainDensityMap.count(bin) == 1) {
-      mainDensityMap.at(bin) += trackDensity;
-    } else {
-      mainDensityMap[bin] = trackDensity;
-    }
+  for (const auto& [bin, trackDensity] : trackDensityMap) {
+    mainDensityMap[bin] += trackDensity;
```

Comment:
Is mainDensityMap[bin] 0 if the bin has not yet been filled?

Reply:
yes it will be default constructed